src/main/java/sx/blah/discord/api/internal/DiscordUtils.java

```diff
@@ -705,17 +706,20 @@ private static IPresence getPresenceFromJSON(GameObject game, String status) {
 	 */
 	public static IEmoji getEmojiFromJSON(IGuild guild, EmojiObject json) {
 		long id = Long.parseUnsignedLong(json.id);
-		IRole[] roles = Arrays.stream(json.roles)
+		List<IRole> roles = Arrays.stream(json.roles)
 				.map(role -> guild.getRoleByID(Long.parseUnsignedLong(role)))
-				.toArray(IRole[]::new);
+				.collect(Collectors.toList());
 
 		EmojiImpl emoji = (EmojiImpl) guild.getEmojiByID(id);
 		if (emoji != null) {
 			emoji.setName(json.name);
-			emoji.setRoles(roles);
+			emoji.setRoles(roles.toArray(new IRole[roles.size()]));
+			return emoji;
 		}
 
-		return new EmojiImpl(id, guild, json.name, Arrays.asList(roles), json.require_colons, json.managed);
+		Cache<IRole> roleCache = new Cache<>((DiscordClientImpl) guild.getClient(), IRole.class);
+		roleCache.putAll(roles);
+		return new EmojiImpl(id, guild, json.name, roleCache, json.require_colons, json.managed);
```

Comment:
Again, only return a new emoji object *if and only if* `emoji == null`

Reply:
This actually is being done, Austin and I discussed.